### PR TITLE
Closes #1207 - Change card breakpoints to match Bootstrap 4 (break sooner)

### LINKED ISSUE #1207: Reassess card deck width settings
## Motivation
The Arizona Bootstrap documentation relates the medium breakpoint with tablet sizes. However, the card deck settings for tablet width set the small breakpoint.

- [Bootstrap docs](https://digital.arizona.edu/arizona-bootstrap/docs/2.0/layout/overview/#responsive-breakpoints)
- [Current tablet card deck settings](https://github.com/az-digital/az_quickstart/blob/46f1fa473b136b7d45e7840bfa31d0a8354b11fa/modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php#L53)

## Proposed resolution
Change the card deck width for tablet settings to set the medium breakpoint. And change the desktop settings to only set the large breakpoint.

Change [desktop settings](https://github.com/az-digital/az_quickstart/blob/46f1fa473b136b7d45e7840bfa31d0a8354b11fa/modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php#L26) to the following:
```
    // Card deck width for desktop.
    $form['card_width'] = [
      '#title' => $this->t('Cards per row on desktop'),
      '#type' => 'select',
      '#options' => [
        'col-lg-12' => $this->t('1'),
        'col-lg-6' => $this->t('2'),
        'col-lg-4' => $this->t('3'),
        'col-lg-3' => $this->t('4'),
      ],
      '#default_value' => $config['card_width'] ?? 'col-lg-4',
      '#description' => $this->t('Choose how many cards appear per row. Additional cards will wrap to a new row. This selection sets the cards per row on desktops with automatic defaults set for tablet and phone. Override cards per row on tablet and phone in Additional options.'),
    ];
```

Change [tablet settings](https://github.com/az-digital/az_quickstart/blob/46f1fa473b136b7d45e7840bfa31d0a8354b11fa/modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php#L53) to the following:
```
    // Card deck width for tablets.
    $form['az_display_settings']['card_width_sm'] = [
      '#title' => $this->t('Cards per row on tablet'),
      '#type' => 'select',
      '#options' => [
        'col-md-12' => $this->t('1'),
        'col-md-6' => $this->t('2'),
        'col-md-4' => $this->t('3'),
        'col-md-3' => $this->t('4'),
      ],
      '#default_value' => $config['az_display_settings']['card_width_sm'] ?? 'col-md-6',
      '#description' => $this->t('Choose how many cards appear per row. Additional cards will wrap to a new row. This selection sets the cards per row on tablets.'),
      '#weight' => 1,
    ];
```

### PR DESCRIPTION
**Pull request by akslay**
## Description
Modifies the card deck break points to match Bootstrap 4. The breakpoints were all bumped up a level in the transition from Bootstrap 3 --> 4, so `-md` should now be `-lg`, and `-sm` should be `-md`. This will help alleviate the issue of cards becoming overly squished at the smaller end of the current "desktop" breakpoint. 

Review site: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-1526.probo.build/

## Related issues
Closes #1207 

## How to test new content (Probo)
- Add some cards to a Page and play with the breakpoints on desktop, tablet, and mobile. 
- Add a sidebar to that same page to ensure it looks reasonable when a sidebar is included.

## How to test existing content (database update on existing site)
- Add the PR diff file as a patch to a local/existing/Pantheon/etc build.
- Run the database update
- Ensure breakpoint selections for card decks update accordingly

### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [x] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [ ] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by joeparsons — APPROVED** [review]

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php [hunk 1]
```diff
@@ -62,10 +62,10 @@ public function buildBehaviorForm(ParagraphInterface $paragraph, array &$form, F
...
-        'col-sm-12' => $this->t('1'),
-        'col-sm-6' => $this->t('2'),
-        'col-sm-4' => $this->t('3'),
-        'col-sm-3' => $this->t('4'),
+        'col-md-12' => $this->t('1'),
+        'col-md-6' => $this->t('2'),
+        'col-md-4' => $this->t('3'),
+        'col-md-3' => $this->t('4'),
       ],
       '#default_value' => $config['az_display_settings']['card_width_sm'] ?? 'col-sm-12',
```
Comment (danahertzberg) [inline modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php:70]:
```suggestion
      '#default_value' => $config['az_display_settings']['card_width_sm'] ?? 'col-md-12',
```

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php [hunk 2]
```diff
@@ -28,10 +28,10 @@ public function buildBehaviorForm(ParagraphInterface $paragraph, array &$form, F
...
-        'col-md-12 col-lg-12' => $this->t('1'),
-        'col-md-6 col-lg-6' => $this->t('2'),
-        'col-md-4 col-lg-4' => $this->t('3'),
-        'col-md-3 col-lg-3' => $this->t('4'),
+        'col-lg-12' => $this->t('1'),
+        'col-lg-6' => $this->t('2'),
+        'col-lg-4' => $this->t('3'),
+        'col-lg-3' => $this->t('4'),
       ],
       '#default_value' => $config['card_width'] ?? 'col-md-4 col-lg-4',
```
Comment (danahertzberg) [inline modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZCardsParagraphBehavior.php:36]:
```suggestion
      '#default_value' => $config['card_width'] ?? 'col-lg-4',
```

**Comment by akslay** [comment]
This will need a database update.

**Comment by joeparsons** [comment]
@tadean Pointed out during our 2022-05-04 meeting that this also needs a database update to update existing paragraphs.

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      ->range($sandbox['current'], $sandbox['current'] + $paragraphs_per_batch)
+      ->execute();
+
+    foreach ($paragraph_ids as $id) {
+      $paragraph = Paragraph::load($id);
+      if (!empty($paragraph)) {
+        $behavior_settings = $paragraph->getAllBehaviorSettings();
+
+        /* Update "desktop" settings. */
+        if (isset($behavior_settings['card_width'])) {
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:43]:
```suggestion
        if (isset($behavior_settings['az_cards_paragraph_behavior']['card_width'])) {
```

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+          switch ($behavior_settings['card_width']) {
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:44]:
```suggestion
          switch ($behavior_settings['az_cards_paragraph_behavior']['card_width']) {
```

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            case 'col-md-12 col-lg-12':
+              $behavior_settings['card_width'] = 'col-lg-12';
+            case 'col-md-6 col-lg-6':
+              $behavior_settings['card_width'] = 'col-lg-6';
+            case 'col-md-4 col-lg-4':
+              $behavior_settings['card_width'] = 'col-lg-4';
+            case 'col-md-3 col-lg-3':
+              $behavior_settings['card_width'] = 'col-lg-3';
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:52]:
```suggestion
            case 'col-md-12 col-lg-12':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-12';
            case 'col-md-6 col-lg-6':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-6';
            case 'col-md-4 col-lg-4':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-4';
            case 'col-md-3 col-lg-3':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-3';
```

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            case 'col-sm-12':
+              $behavior_settings['az_display_settings']['card_width_sm'] = 'col-md-12';
+            case 'col-sm-6':
+              $behavior_settings['az_display_settings']['card_width_sm'] = 'col-md-6';
+            case 'col-sm-4':
+              $behavior_settings['az_display_settings']['card_width_sm'] = 'col-md-4';
+            case 'col-sm-3':
+              $behavior_settings['az_display_settings']['card_width_sm'] = 'col-md-3';
+          }
+        }
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:68]:
```suggestion
        if (isset($behavior_settings['az_cards_paragraph_behavior']['card_width_sm'])) {
          switch ($behavior_settings['az_cards_paragraph_behavior']['card_width_sm']) {
            case 'col-sm-12':
              $behavior_settings['az_cards_paragraph_behavior']['card_width_sm'] = 'col-md-12';
            case 'col-sm-6':
              $behavior_settings['az_cards_paragraph_behavior']['card_width_sm'] = 'col-md-6';
            case 'col-sm-4':
              $behavior_settings['az_cards_paragraph_behavior']['card_width_sm'] = 'col-md-4';
            case 'col-sm-3':
              $behavior_settings['az_cards_paragraph_behavior']['card_width_sm'] = 'col-md-3';
          }
        }
```

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ */
+
+use Drupal\paragraphs\Entity\Paragraph;
+
+/**
+ * Update card deck breakpoints for tablet and desktop media queries.
+ */
+function az_paragraphs_cards_update_9201() {
+
+  if (\Drupal::service('module_handler')->moduleExists('az_paragraphs')) {
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:18]:
```suggestion
```

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+          }
+        }
+
+        $paragraph->setAllBehaviorSettings($behavior_settings);
+        $paragraph->save();
+        $sandbox['updated_count']++;
+      }
+    }
+    $sandbox['current']++;
+  }
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:90]:
```suggestion
```

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * Update card deck breakpoints for tablet and desktop media queries.
+ */
+function az_paragraphs_cards_update_9201() {
+
+  if (\Drupal::service('module_handler')->moduleExists('az_paragraphs')) {
+
+    if (!isset($sandbox['total'])) {
+      $count = \Drupal::entityQuery('paragraph')
+        ->condition('type', 'az_cards')
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:22]:
```suggestion
```
Really any paragraph types can have this behavior enabled on it.

Trying to figure out which ones do, might be complicated, so I think we should just get rid of the condition instead.

File: modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+              break;
+
+            case 'col-md-4 col-lg-4':
+              $behavior_settings['az_cards_paragraph_behavior']['az_display_settings']['card_width'] = 'col-lg-4';
+              break;
+
+            case 'col-md-3 col-lg-3':
+              $behavior_settings['az_cards_paragraph_behavior']['az_display_settings']['card_width'] = 'col-lg-3';
+              break;
+          }
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:60]:
```suggestion
        if (isset($behavior_settings['az_cards_paragraph_behavior']['card_width'])) {
          switch ($behavior_settings['az_cards_paragraph_behavior']['card_width']) {
            case 'col-md-12 col-lg-12':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-12';
              break;

            case 'col-md-6 col-lg-6':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-6';
              break;

            case 'col-md-4 col-lg-4':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-4';
              break;

            case 'col-md-3 col-lg-3':
              $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-3';
              break;
          }
```

Comment (trackleft) [reply modules/custom/az_paragraphs/az_paragraphs_cards/az_paragraphs_cards.install:60]: Here is the shape of the array

```(
    [az_cards_paragraph_behavior] => Array
        (
            [card_width] => col-lg-4
            [card_style] => card
            [az_display_settings] => Array
                (
                    [bottom_spacing] => mb-0
                    [card_width_sm] => col-md-12
                    [card_width_xs] => col-12
                )

        )

)```

**Comment by trackleft** [comment]
I think this is close.

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+ * Install, update and uninstall functions for az_paragraphs_cards install
+ * profile.
+ */
+
+use Drupal\paragraphs\Entity\Paragraph;
+
+/**
+ * Update card deck breakpoints for tablet and desktop media queries.
+ */
+function az_paragraphs_update_9201() {
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:16]: We actually want to add some digits for the Quickstart minor release number here per https://github.com/az-digital/az_quickstart/blob/main/RELEASES.md#update-hook-numbering-convention

01 - if we want to backport to both 2.2.x and 2.1.x
02 - if we only want to backport to 2.2.x
03 - if this should only be included in the 2.3.0 minor release

```suggestion
function az_paragraphs_update_920101() {
```

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 12]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+/**
+ * @file
+ * Az_paragraphs_cards.install.
+ *
+ * Install, update and uninstall functions for az_paragraphs_cards install
```
Comment (akslay) [inline modules/custom/az_paragraphs/az_paragraphs.install:7]:
```suggestion
 * Install, update and uninstall functions for az_paragraphs install
```

**Comment by akslay** [comment]
Looks like the DB update still isn't working even with the changed names to function/comments:
Screen Shot 2022-05-11 at 10 48 31 AM

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+  $paragaph_ids = \Drupal::entityQuery('paragraph')
+    ->range($sandbox['current'], $sandbox['current'] + $paragraphs_per_batch)
+    ->execute();
+
+  foreach ($paragraph_ids as $id) {
+    $paragraph = Paragraph::load($id);
+    if (!empty($paragraph)) {
+      $behavior_settings = $paragraph->getAllBehaviorSettings();
+
+      /* Update "desktop" settings. */
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:38]: Nitpick: c style comments are valid but discouraged.
```suggestion
      // Update "desktop" settings.
```

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+            $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-4';
+            break;
+
+          case 'col-md-3 col-lg-3':
+            $behavior_settings['az_cards_paragraph_behavior']['card_width'] = 'col-lg-3';
+            break;
+        }
+      }
+
+      /* Update "tablet" settings. */
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:59]: Nitpick: c style comments are valid but discouraged.
```suggestion
      // Update "tablet" settings.
```

**Comment by akslay** [comment]
Doesn't look like the DB update is working with the array version either:

Screen Shot 2022-05-11 at 11 48 03 AM

**Comment by akslay** [comment]
DB update is working with the latest commit (aafd0f4). Not working on trying to figure out how to update revisions as well so content using Content Moderation or reverting revisions don't break.

**Review by trackleft — APPROVED** [review]

**Review by joeparsons — APPROVED** [review]

**Review by bberndt-uaz — DISMISSED** [review]

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * @file
+ * Az_paragraphs.install.
+ *
+ * Install, update and uninstall functions for az_paragraphs install
+ * profile.
+ */
+
+use Drupal\paragraphs\Entity\Paragraph;
+
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:12]:
```suggestion
```

File: modules/custom/az_paragraphs/az_paragraphs.install [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  $paragraphs_per_batch = 25;
+
+  $paragraph_ids = \Drupal::entityQuery('paragraph')
+    ->range($sandbox['current'], $sandbox['current'] + $paragraphs_per_batch)
+    ->execute();
+
+  foreach ($paragraph_ids as $id) {
+    $paragraph = Paragraph::load($id);
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.install:34]:
```suggestion
    $paragraph = \Drupal::entityTypeManager()->getStorage('paragraph')->load($id);
```

**Review by mmunro-ltrr — APPROVED** [review]
Did a full local test, running the database update.

**Review by trackleft — APPROVED** [review]

**Review by joeparsons — APPROVED** [review]

**Review by bberndt-uaz — APPROVED** [review]

**Pull request merged**